discovery/gossiper.go

```diff
@@ -1106,6 +1117,30 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(nMsg *networkMsg) []l
 			return nil
 		}
 
+		// If this is a local ChannelUpdate without an AuthProof, it means
+		// it is an update to a channel that is not (yet) supposed to be
+		// announced to the greater network. However, our channel counter
+		// party will need to be given the update, so we'll try sending
+		// the update directly to the remote peer.
+		if !nMsg.isRemote && chanInfo.AuthProof == nil {
+			// Get our peer's public key.
+			var remotePeer *btcec.PublicKey
+			switch {
+			case msg.Flags&lnwire.ChanUpdateDirection == 0:
+				remotePeer = chanInfo.NodeKey2
+			case msg.Flags&lnwire.ChanUpdateDirection == 1:
+				remotePeer = chanInfo.NodeKey1
+			}
+
+			// Send ChannelUpdate directly to remotePeer.
+			// TODO(halseth): make reliable send?
+			if err = d.cfg.SendToPeer(remotePeer, msg); err != nil {
```

Comment:
Atm with this, if we send _just_ the `ChannelUpdate` message, then the other node will reject it as there's a race condition where we might receive the `FundingLocked` then this message before we ever send our _own_ chann ann to our internal gossiper: 
```
2017-12-13 16:16:18.052 [DBG] PEER: Received ChannelUpdate(chain_hash=683e86bd5c6d110d91b94b97137ba6bfe02dbbdb8e3dff722a669b5d69d77af6, short_chan_id=1836184418451456, flag=0, update_time=2017-12-13 16:16:18 -0800 PST) from 127.0.0.1:10019
2017-12-13 16:16:18.052 [TRC] PEER: readMessage from 127.0.0.1:10019: (*lnwire.ChannelUpdate)(0xc42048e9b0)({
 Signature: (*btcec.Signature)(0xc420011140)({
  R: (*big.Int)(0xc420118ea0)(24495463756957618076099430289234328219521735099568467680541926667560037384405),
  S: (*big.Int)(0xc420118ec0)(41256671207089118935243535172289480169661446552925936953485820156560683126184)
 }),
 ChainHash: (chainhash.Hash) (len=32 cap=32) 683e86bd5c6d110d91b94b97137ba6bfe02dbbdb8e3dff722a669b5d69d77af6,
 ShortChannelID: (lnwire.ShortChannelID) {
  BlockHeight: (uint32) 1670,
  TxIndex: (uint32) 1,
  TxPosition: (uint16) 0
 },
 Timestamp: (uint32) 1513210578,
 Flags: (lnwire.ChanUpdateFlag) 0,
 TimeLockDelta: (uint16) 144,
 HtlcMinimumMsat: (lnwire.MilliSatoshi) 0 mSAT,
 BaseFee: (uint32) 1000,
 FeeRate: (uint32) 1
})

2017-12-13 16:16:18.053 [ERR] DISC: unable to validate channel update short_chan_id=1836184418451456: edge not found
```

So one side doesn't see the other's edge:
```
⛰i  lncli describegraph
{
    "nodes": [
        {
            "last_update": 0,
            "pub_key": "02cc27b8ef708b13a6832f1c4f3d8d1207851cda53f5be041f5b1e2ec1afeb755c",
            "alias": "",
            "addresses": [
            ],
            "color": "#000000"
        },
        {
            "last_update": 1513209865,
            "pub_key": "03d381f8cd62e79e40f2a0f96caa0f2e9ee1f6526e80a05c4907dcdf8627f0974e",
            "alias": "03d381f8cd62e79e40f2\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000",
            "addresses": [
            ],
            "color": "#3399ff"
        }
    ],
    "edges": [
        {
            "channel_id": "1836184418451456",
            "chan_point": "1592c661827f01270f3e309e82770150eb40fc1df4d4357c1d89c7bc5cf71cef:0",
            "last_update": 1513210579,
            "node1_pub": "02cc27b8ef708b13a6832f1c4f3d8d1207851cda53f5be041f5b1e2ec1afeb755c",
            "node2_pub": "03d381f8cd62e79e40f2a0f96caa0f2e9ee1f6526e80a05c4907dcdf8627f0974e",
            "capacity": "1500000",
            "node1_policy": null,
            "node2_policy": {
                "time_lock_delta": 144,
                "min_htlc": "0",
                "fee_base_msat": "1000",
                "fee_rate_milli_msat": "1"
            }
        }
    ]
}
```

This means that they wouldn't be able to craft a routing hint to route to the private channel.

Reply:
woops, this was a big oversight on my part!

I pushed a commit + tests now, that takes care of this case. It stores the `ChannelUpdate` for the unknown channel in a map, and will reprocess it later, after the local `ChannelAnnouncement` is processed. At the moment this is just a map in memory, not sure if it is needed/worth it at this point to make it persistent. @aakselrod 

If this looks okay, I will squash and rebase this branch, getting it ready for merge 💪